titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeOpportunisticResourceProvider.java

```diff
@@ -176,19 +193,15 @@ private CodeInvariants codeInvariants() {
      */
     @Override
     public Map<String, OpportunisticCpuAvailability> getOpportunisticCpus() {
-        Map<String, OpportunisticCpuAvailability> cpuAvailability = informer.getIndexer().list().stream().collect(Collectors.toMap(
+        return informer.getIndexer().list().stream().collect(Collectors.toMap(
                 V1OpportunisticResource::getInstanceId,
                 resource -> new OpportunisticCpuAvailability(resource.getUid(), resource.getEnd(), resource.getCpus()),
                 BinaryOperator.maxBy(EXPIRES_AT_COMPARATOR)
         ));
-        lastCpuSupply.set(cpuAvailability.values().stream()
```

Comment:
Since this method is being called many times for scheduling, I think it would be better if you create an [active gauge](http://netflix.github.io/spectator/en/latest/intro/gauge/#active-gauges) in the constructor and then set it in this method so that the min/max is correct. The downside to using the `PolledMeter` is that you don't know the rate at which it is being polled and the min/max values might not be accurate.

Reply:
I did something similar to what you are suggesting before (each time this method gets called, update an `AtomicLong` with the current value, then install a `PolledMeter` on the `AtomicLong`).

The problem is that the rate in which this method gets called is orders of magnitude higher than the polling frequency (by `PolledMeter`). We know the polling frequency because it's configurable through `RegistryConfig#gaugePollingFrequency()` (default of 10s). Not  computing the metric inline with all calls (including the ones from scheduling hooks) ends up being way cheaper.

As far of min/max, I looked into the code of the active `Gauge`, and it doesn't track min/max for each step, it only keeps the last computed value. To have better control over we would need to use `distributionSummary` or `maxGauge`.